autonat/autonat-v2.md

```diff
@@ -19,71 +19,89 @@ Interest Group: [@marten-seemann], [@marcopolo], [@mxinden]
 
-`autonat v2` allows nodes to determine reachability for individual addresses. 
-Using `autonat v2` nodes can build an address pipeline where they can test 
-individual addresses discovered by different sources like identify, upnp 
-mappings, circuit addresses etc for reachability.
+In `autonat v2` client sends a priority ordered list of addresses. On receiving
+this list the server dials the first address on the list that it is capable of
+dialing. `autonat v2` allows nodes to determine reachability for individual
+addresses. Using `autonat v2` nodes can build an address pipeline where they can
+test individual addresses discovered by different sources like identify, upnp
+mappings, circuit addresses etc for reachability. Having a priority ordered list
+of addresses provides the ability to verify low priority addresses.
+Implementations can generate low priority address guesses and add them to
+requests for high priority addresses as a nice to have. 
 
 Compared to `autonat v1` there are two major differences
-1. `autonat v1` allowed testing reachability for the node. `autonat v2` allows for testing reachability for an individual address
-2. `autonat v2` provides a mechanism for nodes to verify whether the peer 
+1. `autonat v1` allowed testing reachability for the node. `autonat v2` allows
+testing reachability for an individual address
+2. `autonat v2` provides a mechanism for nodes to verify whether the peer
 actually successfully dialled an address.
 
 
 ## AutoNAT V2 Protocol
 
-A node wishing to determine reachability of a particular adddress sends a 
-`DialRequest` message to peer B on a stream with protocol ID
-`/libp2p/autonat/2.0.0/dial-request`. This `DialRequest` message has the 
-address and a uint64 nonce. 
+A node wishing to determine reachability of its adddresses sends a `DialRequest`
+message to a peer on a stream with protocol ID
+`/libp2p/autonat/2.0.0/dial-request`. 
 
-Upon receiving this message the peer attempts to dial the address and opens a 
-stream with  Protocol ID `/libp2p/autonat/2.0.0/dial-attempt` and sends a 
-`DialAttempt` message with the nonce received in the `DialRequest`. The peer 
-MUST NOT dial any address other than the address provided in the `DialRequest` 
-message.
+This `DialRequest` message has a list of `AddressDialRequest`s. Each item in
+this list contains an address and a fixed64 nonce. The list is ordered in
+descending order of priority for verfication.
 
-Upon completion of the dial attempt, the peer sends a `DialResponse` to the 
-initiator node on the `/libp2p/autonat/2.0.0/dial-request` stream.
+Upon receiving this message the peer attempts to dial the first address from the
+list that it is capable of dialing. It dials this address, opens a stream with
+Protocol ID `/libp2p/autonat/2.0.0/dial-attempt` and sends a `DialAttempt`
+message with the nonce received in the corresponding `AddressDialRequest`. The
+peer MUST dial the first address in the list it is capable of dialing. The peer
```

Comment:
The logic is correct, but we shouldn't use a MUST here. The peer can still refuse to dial, e.g. due to rate-limiting.

Reply:
fixed. I removed MUST and added this requirement to MUST NOT phrase.